External traffic policy (#728)
kyma-project/istio: docs/user/tutorials/01-00-x-forwarded-for-header.md

Reviewer: ```
> [!NOTE]
>  The X-Forwarded-For header is only supported on AWS clusters.
```
This needs to be removed.

Also it's no longer a Prerequisite to have an AWS cluster, since it should also work on GCP and Azure

Many applications need to know the client IP address of an originating request to behave properly. Usual use cases include workloads that require the client IP address to restrict their access. The ability to provide client attributes to services has long been a staple of reverse proxies, which use the X-Forwarded-For (XFF) header to forward client attributes to destination workloads. For more information on XFF, see 
the [IETF’s RFC documentation](https://datatracker.ietf.org/doc/html/rfc7239) and [Envoy documentation](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#x-forwarded-for).

## Prerequisites

> * An AWS cluster

Reviewer 2: ```suggestion
* An AWS or GCP cluster
```

  ```

#### **Kyma Dashboard**
1. Navigate to **Cluster Details** and select **Modify Modules**.
2. Choose the Istio module and select **Edit**.
3. In the `General` section, add the number of trusted proxies.
4. Select **Save**.
<!-- tabs:end -->

### Configure Gateway External Traffic Policy in the Istio Custom Resource (GCP and Azure only)

If you are using a GCP or Azure cluster, you must also set the **gatewayExternalTrafficPolicy** to `Local` in order to include the client's IP address in the XFF header. Skip this step if you're using a different cloud service provider.

> 

Reviewer 2: ```suggestion

For production Deployments, it is strongly recommended to deploy an Ingress Gateway Pod to multiple nodes if you enable `externalTrafficPolicy: Local`. Otherwise, this creates a situation where only nodes with an active Ingress Gateway Pod are able to accept and distribute incoming NLB traffic to the rest of the cluster, creating potential ingress traffic bottlenecks and reduced internal load balancing capability, or even complete loss of ingress traffic incoming to the cluster if the subset of nodes with Ingress Gateway Pods goes down. See the source IP for Services with `Type=NodePort` for more information. For reference, see Istio [Network Load Balancer](https://istio.io/latest/docs/tasks/security/authorization/authz-ingress/#network) documentation.
Default Kyma Istio installation profile configures **PodAntiAffinity** to ensure that Ingress Gateway Pods are evenly spread across all Nodes. This guarantees that the above requirement is satisfied if your IngressGateway autoscaling configuration **minReplicas** is at least equal to the number of Nodes. You can configure autoscaling options in the Istio custom resource using **spec.config.components.ingressGateway.k8s.hpaSpec.minReplicas**.

> [!WARNING]
> Deploy an Ingress Gateway Pod to multiple nodes if you enable `externalTrafficPolicy: Local` in production Deployments.
```

> [!TIP]
> While using GCP or Azure, you can find your load balancer's IP address in the field **status.loadBalancer.ingress** of the `ingress-gateway` Service.

Add **gatewayExternalTrafficPolicy** to the Istio custom resource:

<!-- tabs:start -->
#### **kubectl**
Run the following command:

  ```bash
  kubectl patch istios/default -n kyma-system --type merge -p '{"spec":{"config":{"gatewayExternalTrafficPolicy": "Local"}}}'
  ```

> [!WARNING]
> For production deployments it is strongly recommended to deploy an ingress gateway pod to multiple nodes if you enable externalTrafficPolicy: Local. Otherwise, this creates a situation where only nodes with an active ingress gateway pod will be able to accept and distribute incoming NLB traffic to the rest of the cluster, creating potential ingress traffic bottlenecks and reduced internal load balancing capability, or even complete loss of ingress traffic to the cluster if the subset of nodes with ingress gateway pods go down. See Source IP for Services with Type=NodePort for more information. See reference Istio documentation: https://istio.io/latest/docs/tasks/security/authorization/authz-ingress/#network

> > Default Kyma Istio installation profile configures PodAntiAffinity to ensure that IngressGateway pods are evenly spread against all Nodes. This guarantees that if your IngressGateway autoscaling configuration `minReplicas` is at least equal to the number of Nodes, the above requirement will be satisfied. You can configure autoscaling options in Istio Custom Resource using `spec.config.components.ingressGateway.k8s.hpaSpec.minReplicas`.

Reviewer 2: ```suggestion
> Default Kyma Istio installation profile configures **PodAntiAffinity** to ensure that Ingress Gateway Pods are evenly spread across all Nodes. This guarantees that the above requirement is satisfied if your IngressGateway autoscaling configuration **minReplicas** is at least equal to the number of Nodes. You can configure autoscaling options in the Istio custom resource using **spec.config.components.ingressGateway.k8s.hpaSpec.minReplicas**.
```

4. Select **Save**.
<!-- tabs:end -->

### Configure Gateway External Traffic Policy in the Istio Custom Resource (GCP and Azure only)

If you are using a GCP or Azure cluster, you must also set the **gatewayExternalTrafficPolicy** to `Local` in order to include the client's IP address in the XFF header. Skip this step if you're using a different cloud service provider.

For production Deployments, it is strongly recommended to deploy an Ingress Gateway Pod to multiple nodes if you enable `externalTrafficPolicy: Local`. Otherwise, this creates a situation where only nodes with an active Ingress Gateway Pod are able to accept and distribute incoming NLB traffic to the rest of the cluster, creating potential ingress traffic bottlenecks and reduced internal load balancing capability, or even complete loss of ingress traffic incoming to the cluster if the subset of nodes with Ingress Gateway Pods goes down. See the source IP for Services with `Type=NodePort` for more information. For reference, see Istio [Network Load Balancer](https://istio.io/latest/docs/tasks/security/authorization/authz-ingress/#network) documentation.

> Default Kyma Istio installation profile configures **PodAntiAffinity** to ensure that Ingress Gateway Pods are evenly spread across all Nodes. This guarantees that the above requirement is satisfied if your IngressGateway autoscaling configuration **minReplicas** is at least equal to the number of Nodes. You can configure autoscaling options in the Istio custom resource using **spec.config.components.ingressGateway.k8s.hpaSpec.minReplicas**.

Reviewer 2: ```suggestion

Default Kyma Istio installation profile configures **PodAntiAffinity** to ensure that Ingress Gateway Pods are evenly spread across all Nodes. This guarantees that the above requirement is satisfied if your IngressGateway autoscaling configuration **minReplicas** is at least equal to the number of Nodes. You can configure autoscaling options in the Istio custom resource using **spec.config.components.ingressGateway.k8s.hpaSpec.minReplicas**.
```

> [!WARNING]
> Deploy an Ingress Gateway Pod to multiple nodes if you enable `externalTrafficPolicy: Local` in production Deployments.

> > [!TIP]
> > While using GCP or Azure, you can find your load balancer's IP address in the field **status.loadBalancer.ingress** of the `ingress-gateway` Service.

Reviewer 2: ```suggestion

> [!TIP]
> While using GCP or Azure, you can find your load balancer's IP address in the field **status.loadBalancer.ingress** of the `ingress-gateway` Service.
```

Default Kyma Istio installation profile configures **PodAntiAffinity** to ensure that Ingress Gateway Pods are evenly spread across all Nodes. This guarantees that the above requirement is satisfied if your IngressGateway autoscaling configuration **minReplicas** is at least equal to the number of Nodes. You can configure autoscaling options in the Istio custom resource using **spec.config.components.ingressGateway.k8s.hpaSpec.minReplicas**.

> [!WARNING]
> Deploy an Ingress Gateway Pod to multiple nodes if you enable `externalTrafficPolicy: Local` in production Deployments.
> [!TIP]
> While using GCP or Azure, you can find your load balancer's IP address in the field **status.loadBalancer.ingress** of the `ingress-gateway` Service.

Add **gatewayExternalTrafficPolicy** to the Istio custom resource:

<!-- tabs:start -->
#### **kubectl**
Run the following command:

  ```bash
  kubectl patch istios/default -n kyma-system --type merge -p '{"spec":{"config":{"gatewayExternalTrafficPolicy": "Local"}}}'
  ```

> > [!WARNING]
> > For production deployments it is strongly recommended to deploy an ingress gateway pod to multiple nodes if you enable externalTrafficPolicy: Local. Otherwise, this creates a situation where only nodes with an active ingress gateway pod will be able to accept and distribute incoming NLB traffic to the rest of the cluster, creating potential ingress traffic bottlenecks and reduced internal load balancing capability, or even complete loss of ingress traffic to the cluster if the subset of nodes with ingress gateway pods go down. See Source IP for Services with Type=NodePort for more information. See reference Istio documentation: https://istio.io/latest/docs/tasks/security/authorization/authz-ingress/#network
> > Default Kyma Istio installation profile configures PodAntiAffinity to ensure that IngressGateway pods are evenly spread against all Nodes. This guarantees that if your IngressGateway autoscaling configuration `minReplicas` is at least equal to the number of Nodes, the above requirement will be satisfied. You can configure autoscaling options in Istio Custom Resource using `spec.config.components.ingressGateway.k8s.hpaSpec.minReplicas`.

Reviewer 2: ```suggestion
```

  ```

#### **Kyma Dashboard**
1. Navigate to Cluster Details and select Modify Modules.
2. Choose the Istio module and select Edit.
3. In the General section, add the number of trusted proxies.
4. Select Save.
<!-- tabs:end -->

> ### Configure the externalTrafficPolicy in the Istio Custom Resource (GCP and Azure only)

Reviewer 2: ```suggestion
### Configure the External Traffic Policy in the Istio Custom Resource (GCP and Azure Only)
```

Author: The k8s field name is [externalTrafficPolicy](https://kubernetes.io/docs/tasks/access-application-cluster/create-external-load-balancer/#preserving-the-client-source-ip), as so I would not like it to be separated

Author: OK, now I see that't a busola tutorial :D
Maybe then we could use the full name here as well (Gateway external traffic policy)?

Reviewer 2: There are Busola and kubectl instructions, but I agree that Gateway External Traffic Policy would be more universal

> To make XFF header contain client's IP on GCP and Azure, it's required to set gatewayExternalTrafficPolicy to `Local`

Reviewer 2: ```suggestion
You can configure external traffic policy only in GCP and Azure clusters. To include the client's IP in the XFF header, set **gatewayExternalTrafficPolicy** to `Local`.
```

Reviewer 3: We can configure it on every cloud provider. But to have XFF header forwarding the client's IP on Azure and GCP you have to set gatewayExternalTrafficPolicy to Local. On AWS it does not have any effect regarding the XFF header content.

Reviewer 2: ahh, okay 
sorry, I didn't get this

Reviewer 2: I've edited the suggestion

> Hint: While using GCP or Azure you can find your load balancer's IP address in the ingress-gateway Service's status.loadBalancer.ingress field.

Reviewer 2: ```suggestion
> [!TIP]
> While using GCP or Azure, you can find your load balancer's IP address in the field **status.loadBalancer.ingress** of the `ingress-gateway` Service.
```

> Add **gatewayExternalTrafficPolicy** to the Istio custom resource:

Reviewer 2: ```suggestion
Set **gatewayExternalTrafficPolicy** to `Local` in the Istio custom resource:
```

<!-- tabs:start -->
#### **kubectl**
Run the following command:

  ```bash
  kubectl patch istios/default -n kyma-system --type merge -p '{"spec":{"config":{"gatewayExternalTrafficPolicy": "Local"}}}'
  ```

#### **Kyma Dashboard**

> 1. Navigate to Cluster Details and select Modify Modules.

Reviewer 2: ```suggestion
1. Navigate to **Cluster Details** and select **Modify Modules**.
```

> 2. Choose the Istio module and select Edit.

Reviewer 2: ```suggestion
2. Choose the Istio module and select **Edit**.
```

> 3. In the General section, select `Local` in gatewayExternalTrafficPolicy field.

Reviewer 2: ```suggestion
3. In the `General` section, set the external traffic policy for Istio Ingress Gateway to `Local`.
```

Author: The ui field name is `gatewayExternalTrafficPolicy`, as so it might be confusing for the user if we don't use the exact name here

Reviewer 2: But it's not visible in Busola, right? I'd use the same name here as in Busola. So, "Gateway external traffic policy" if we want to stick with it

Author: Right, here it makes sense as this an UI tutorial, my bad

> 4. Select Save.

Reviewer 2: ```suggestion
4. Select **Save**.
```

  ```

#### **Kyma Dashboard**
1. Navigate to Cluster Details and select Modify Modules.
2. Choose the Istio module and select Edit.
3. In the General section, add the number of trusted proxies.
4. Select Save.
<!-- tabs:end -->

> ### Configure the externalTrafficPolicy in the Istio Custom Resource

Reviewer: I think we should scope this to be only relevant for AWS and GCP.

***GCP and Azure***

GCP and Azure require to set gatewayExternalTrafficPolicy to `Local`

Add **gatewayExternalTrafficPolicy** to the Istio custom resource:

<!-- tabs:start -->
#### **kubectl**
Run the following command:

  ```bash
  kubectl patch istios/default -n kyma-system --type merge -p '{"spec":{"config":{"gatewayExternalTrafficPolicy": "Local"}}}'
  ```

#### **Kyma Dashboard**
1. Navigate to Cluster Details and select Modify Modules.
2. Choose the Istio module and select Edit.
3. In the General section, select `Local` in gatewayExternalTrafficPolicy field.
4. Select Save.
<!-- tabs:end -->

> ***AWS***

Reviewer: I would not add this information. The user should not care about it. 
For AWS this is working out of the box.
